src/ray/util/logging.h

```diff
@@ -126,6 +126,17 @@ enum class RayLogLevel {
 
 #endif  // NDEBUG
 
+#define RAY_CHECK_OP(left, op, right)                                         \
+  if (const auto &[_left_, _right_] = std::make_tuple((left), (right)); true) \
```

Comment:
what's this for?

Reply:
I'm not sure if I understand your questions. This is C++ 17 structured binding declaration. https://stackoverflow.com/questions/46156398/structured-binding-initializer-inside-if-statement-does-not-compile

Reply:
+1 is the extra copy necessary?

Reply:
Not necessary but I don't know what's a good way to avoid it. Let me know if you have any!

I tried
```
#define RAY_CHECK_OP(left, op, right)                                         
  if (const auto &_left_ = (left), &_right_ = (right); true)               
  RAY_CHECK((_left_ op _right_)) << " " << _left_ << " vs " << _right_
```

The issue is that compiler needs to deduce the same type for `left` and `right` and without an explicit `static_cast` it somethings failed to do that:
```
 src/ray/raylet/wait_manager.cc:26:3: error: 'auto' deduced as 'unsigned long long' in declaration of '_left_' and deduced as 'unsigned long' in declaration of '_right_'
      RAY_CHECK_LE(num_required_objects, object_ids.size());
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

We can let users manually do `static_cast`, the downside is that this is not user friendly :)

Reply:
I think I figured out a way to avoid the extra copy, please take another look.

Reply:
What are the cases when the following does not work?
```
#define RAY_CHECK_OP(left, op, right)        \
  RAY_CHECK(((left) op (right))) << " " << (left) << " vs " << (right)
```

Reply:
If left is a function call, we don't want it to be executed twice: RAY_CHECK_EQ(Func(), 1);

Reply:
Makes sense.